container/lxd/lxd.go

```diff
@@ -176,78 +420,62 @@ func (manager *containerManager) CreateContainer(
 	}
 
-	spec := lxdclient.InstanceSpec{
-		Name:     name,
-		Image:    imageName,
-		Metadata: metadata,
-		Devices:  nics,
-		Profiles: profiles,
+	spec := api.ContainersPost{
+		Name: name,
+		ContainerPut: api.ContainerPut{
+			Profiles: profiles,
+			Devices:  nics,
+			Config:   metadata,
+		},
 	}
 
-	logger.Infof("starting instance %q (image %q)...", spec.Name, spec.Image)
-	callback(status.Provisioning, "Starting container", nil)
-	_, err = manager.client.AddInstance(spec)
+	logger.Infof("starting instance %q (image %q)...", spec.Name, imageName)
+	callback(status.Provisioning, "Creating container", nil)
+	op, err := manager.server.CreateContainerFromImage(imageServer, *image, spec)
 	if err != nil {
-		return
-	}
-
-	callback(status.Running, "Container started", nil)
-	inst = &lxdInstance{name, manager.client}
-	return
-}
-
-func (manager *containerManager) DestroyContainer(id instance.Id) error {
-	if manager.client == nil {
-		var err error
-		manager.client, err = ConnectLocal()
-		if err != nil {
-			return err
-		}
+		logger.Errorf("CreateContainer failed with %s", err)
+		return "", errors.Trace(err)
 	}
-	return errors.Trace(manager.client.RemoveInstances(manager.namespace.Prefix(), string(id)))
-}
 
-func (manager *containerManager) ListContainers() (result []instance.Instance, err error) {
-	result = []instance.Instance{}
-	if manager.client == nil {
-		manager.client, err = ConnectLocal()
-		if err != nil {
+	progress := func(op api.Operation) {
+		if op.Metadata == nil {
 			return
 		}
+		for _, key := range []string{"fs_progress", "download_progress"} {
+			value, ok := op.Metadata[key]
+			if ok {
+				callback(status.Provisioning, fmt.Sprintf("Retrieving image: %s", value.(string)), nil)
+				return
+			}
+		}
 	}
-
-	lxdInstances, err := manager.client.Instances(manager.namespace.Prefix())
+	_, err = op.AddHandler(progress)
 	if err != nil {
-		return
+		return "", errors.Trace(err)
 	}
 
-	for _, i := range lxdInstances {
-		result = append(result, &lxdInstance{i.Name, manager.client})
+	op.Wait()
+	opInfo, err := op.GetTarget()
+	if err != nil {
+		return "", errors.Trace(err)
 	}
-
-	return
-}
-
-func (manager *containerManager) IsInitialized() bool {
-	if manager.client != nil {
-		return true
+	if opInfo.StatusCode != api.Success {
+		return "", fmt.Errorf("LXD error: %s", opInfo.Err)
 	}
-
-	// NewClient does a roundtrip to the server to make sure it understands
-	// the versions, so all we need to do is connect above and we're done.
-	var err error
-	manager.client, err = ConnectLocal()
-	return err == nil
+	return name, nil
 }
 
-// HasLXDSupport returns false when this juju binary was not built with LXD
-// support (i.e. it was built on a golang version < 1.2
-func HasLXDSupport() bool {
-	return true
+func (manager *containerManager) removeInstance(name string) error {
+	op, err := manager.server.DeleteContainer(name)
+	if err != nil {
+		return err
+	}
+	err = op.Wait()
```

Comment:
Will this wait forever?  Is there a timeout defined somewhere else?

Reply:
Theoretically - yes. Practically - LXD will timeout internally and return an error sooner.